File: 242.isAnagram.c
#include "shdstd.h"

#define MA(c) ((c) - 'a')

bool isAnagram(char *src, char *trg)
{
	ssize_t src_len = strlen(src);
	ssize_t trg_len = strlen(trg);

	if (src_len != trg_len) {
		return false;
	}

	int src_freq[26];
	int trg_freq[26];

	for (int i = 0; i < src_len; ++i) {
		int src_ch = src[i];
		int trg_ch = trg[i];
		++src_freq[MA(src_ch)];
		++trg_freq[MA(trg_ch)];
	}

	for (int i = 0; i < 26; ++i) {
		if (src_freq[i] != trg_freq[i]) {
			return false;
		}
	}

	return true;
}

File: 242.isAnagram.go
package main

func code(ch byte) int {
	return int(ch - 'a')
}

func isAnagram(src string, trg string) bool {
	src_len := len(src)
	trg_len := len(trg)

	if src_len != trg_len {
		return false
	}

	src_freq := [26]int{}
	trg_freq := [26]int{}

	for i := 0; i < src_len; i++ {
		src_ch := src[i]
		trg_ch := trg[i]

		src_freq[code(src_ch)]++
		trg_freq[code(trg_ch)]++
	}

	for i := 0; i < 26; i++ {
		if src_freq[i] != trg_freq[i] {
			return false
		}
	}

	return true
}
